tests/test_prolongation.cpp

```diff
@@ -0,0 +1,297 @@
+
+    gmgpolar test_p;
+    create_grid(test_p);
+
+    level& p_level = *(test_p.v_level[0]);
+    int ctheta_int = test_p.v_level[1]->ntheta_int; //number of coarse nodes in theta direction
+
+    p_level.m  = test_p.v_level[0]->nr * test_p.v_level[0]->ntheta; //fine grid size
+    p_level.mc = test_p.v_level[1]->nr * test_p.v_level[1]->ntheta; //coarser grid size
+
+    std::vector<double> u_test(p_level.mc);
+    for (int z = 0; z < p_level.mc; z++) {
+        u_test[z] =
+            1 - z +
+            pow(PI,
+                -z * z); //constructing arbitrary grid-function on coarse level to test our prolongation operator with.
+    }
+
+    std::vector<double> sol =
+        p_level.apply_prolongation_bi(u_test); //apply prolongation operator on arbitrary grid-function
+
+    for (int j = 0; j < p_level.nr_int + 1; j++) {
+        for (int i = 0; i < p_level.ntheta_int; i++) {
+            if (j % 2 == 0 && i % 2 == 0) { //testing coarse node injection
+                EXPECT_EQ(u_test[(j / 2) * ctheta_int + (i / 2)], sol[j * p_level.ntheta_int + i])
+                    << "coarse node injection is failing";
+            }
+            else if (i % 2 != 0 && j % 2 == 0) {
+            // as numbering in angle (theta_i) is odd, we have a fine node with
+            // coarse neighbors at (r_j, theta_i - k_{i-1}) and (r_j, theta_i + k_i)
+
+                double k_qm1 = p_level.theta[i] - p_level.theta[i - 1]; //calculate k_{q-1}
+                double k_q   = (i < p_level.ntheta_int - 1) ? p_level.theta[i + 1] - p_level.theta[i]
+                                                          : 2 * PI - p_level.theta[i]; //k_q
+
+                int i1 = (j / 2) * ctheta_int + (i - 1) / 2; //bottom coarse node in theta
+                int i2 = (i < p_level.ntheta_int - 1) ? i1 + 1 : (j / 2) * ctheta_int; //top coarse node in theta
+
+                double val = (k_q * u_test[i1] + k_qm1 * u_test[i2]);
+
+                EXPECT_NEAR(val, (k_q + k_qm1) * sol[j * p_level.ntheta_int + i], 1e-8) //compare values as in the paper
+                    << "Bilinear Prolongation fails for Index (r,theta) : (" + std::to_string(j) + "," +
+                           std::to_string(i) + ")";
+                ;
+            }
+            else if (i % 2 == 0 && j % 2 != 0) {
+            // as numbering in radius (r_j) is odd, we have a fine node with
+            // coarse neighbors at (r_j - h_{j-1}, theta_i) and (r_j+h_j, theta_i )
+                double h_pm1 = p_level.r[j] - p_level.r[j - 1]; //h_{p-1}
+                double h_p   = p_level.r[j + 1] - p_level.r[j]; //h_p
+
+                double v1 = u_test[(j - 1) / 2 * ctheta_int + (i / 2)]; // left coarse node in r
+                double v2 = u_test[(j + 1) / 2 * ctheta_int + (i / 2)]; // right coarse node in r
+
+                double val = (h_p * v1 + h_pm1 * v2);
+
+                EXPECT_NEAR(val, (h_p + h_pm1) * sol[j * p_level.ntheta_int + i], 1e-8) //compare values
```

Comment:
```suggestion
                EXPECT_NEAR(val, (h_p + h_pm1) * sol[j * p_level.ntheta_int + i], 1e-12) //compare values
```
Can we make this succeed with 1e-12 too?